### PR TITLE
Integrated child tree proof

### PR DESCRIPTION
**Pull request by mustermeiszer**
Used in order to compare, my changes to the previous stage of the crowdloan branch.
This PR does change the structure of the crowdloan to acknowledge the usage of the internal verification abilities for child-tree-keys.

**Changes**
* Remove contributor registration
* Ensure reward is called after init of module
* Integrate TrieIndex storage
* Integrate proof verification

### CODE REVIEW

**Comment by mustermeiszer** [comment]
Once this is merged I can rebase the reward branch, merge #342 and start testing.

**Comment by mustermeiszer** [comment]
Turned back to WIP due to the changes of the `validate_unsigned` that will be necessary.

**Comment by mustermeiszer** [comment]
Please review `validate_unsigned`,  the rest of the logic did not change except for the removal of the tick increase. Tests will come once the proofs have been extracted.

File: pallets/crowdloan-claim/src/tests.rs [hunk 1]
```diff
@@ -55,38 +55,44 @@ use frame_support::{
...
+#[test]
+fn test_invalid_claim_wrong_relayaccount() {
+    // TODO: init module, valid proof, valid amount, wrong relay account
 }
- 
\ No newline at end of file
+
+#[test]
+fn test_invalid_claim_mod_not_initalized() {
+    // TODO: claim wohtout init of module
```
Comment (vedhavyas) [inline pallets/crowdloan-claim/src/tests.rs:95]: do we wait until these tests are added ?

Comment (mustermeiszer) [reply pallets/crowdloan-claim/src/tests.rs:95]: No, I need to extract proofs and so on from the chain first, which is currently slightly annoying, as the current polkadot API version does not yet contain the `getChildReadProof`.

File: pallets/crowdloan-claim/src/lib.rs [hunk 2]
```diff
@@ -417,120 +357,68 @@ pub mod pallet {
...
-        ///
-        /// A registered contributor proves her/his identity to the parachain,
-        /// given a relay chain account id, a parachain account id and a
-        /// signature, in order to prove her/his native identity.
-        /// \[relaychain_account_id, parachain_account_id\]
-        ContributorRegistered(T::RelayChainAccountId, ParachainAccountIdOf<T>),
-
         /// Event triggered when a reward has already been processed.
         /// \[who, amount\]
         ClaimAlreadyProcessed(T::RelayChainAccountId, ContributionAmountOf<T>),
```
Comment (vedhavyas) [inline pallets/crowdloan-claim/src/lib.rs:362]: Do we need this event? Can we error out when the contributor already claimd the reward ?

File: pallets/crowdloan-claim/src/lib.rs [hunk 3]
```diff
@@ -417,120 +357,68 @@ pub mod pallet {
...
-    /// so, the signed [`register_contributor`] transaction must be first called. 
+    /// This is needed in order to build the correct keys for proof check.
     #[pallet::storage]
-	#[pallet::getter(fn claimed_relay_chain_ids)]
-	pub type RegisteredContributors<T: Config> = StorageMap<
-        _, 
-        Blake2_128Concat, 
-        T::RelayChainAccountId, ()
-    >;
+    #[pallet::getter(fn trie_index)]
```
Comment (vedhavyas) [inline pallets/crowdloan-claim/src/lib.rs:395]: maybe `crowdloan_trie_index`?

File: pallets/crowdloan-claim/src/lib.rs [hunk 4]
```diff
@@ -417,120 +357,68 @@ pub mod pallet {
...
-	///
-	/// To prevent spam (or Denial of Service - DOS) of unsigned claim transactions 
-    /// on the network, claim we only allow one transaction every `T::ClaimTransactionInterval`
-	/// blocks. This storage entry defines the next tick when a new claim transaction
-    /// can be performed. It acts as a kind of throttling mechanism for (feeless)
-    /// claim transactions to be executed at a definite beat.
-    #[pallet::storage]
-	#[pallet::getter(fn next_unsigned_transaction_at)]
-	pub(super) type ClaimTransactionTick<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+    #[pallet::getter(fn claims_processed)]
```
Comment (vedhavyas) [inline pallets/crowdloan-claim/src/lib.rs:400]: `processed_claims`

File: pallets/crowdloan-claim/src/lib.rs [hunk 5]
```diff
@@ -566,195 +450,234 @@ pub mod pallet {
...
-            contribution_proof: [proof_type_here]
+            contribution_proof: Vec<Vec<u8>>,
         ) -> DispatchResultWithPostInfo {
-            // Ensures that this function can only be called via an unsigned transaction			
+            // Ensures that this function can only be called via an unsigned transaction
             ensure_none(origin)?;
 
+            // Ensure that the module has been initialized before calling this
+            ensure!(
+                <Contributions<T>>::get() != None,
```
Comment (vedhavyas) [inline pallets/crowdloan-claim/src/lib.rs:485]: maybe use `is_some`. I think that is more readable no ?

File: pallets/crowdloan-claim/src/lib.rs [hunk 6]
```diff
@@ -566,195 +450,234 @@ pub mod pallet {
...
-            ensure!(<Contributions<T>>::get() == None, Error::<T>::PalletAlreadyInitialized);
+            ensure!(
+                Self::ensure_administrator(origin) == Ok(()),
+                Error::<T>::MustBeAdministrator
+            );
+
+            // Ensure that the pallet has not already been initialized
+            // TODO: Should we change this? This would basically allow the pallet to be used exactly once, without an RuntimeUpgrade.
+            ensure!(
+                <Contributions<T>>::get() == None,
```
Comment (vedhavyas) [inline pallets/crowdloan-claim/src/lib.rs:560]: `is_none` maybe ?

File: pallets/crowdloan-claim/src/lib.rs [hunk 7]
```diff
@@ -566,195 +450,234 @@ pub mod pallet {
...
+            contributions: ChildTrieRootHashOf<T>,
+            index: TrieIndex,
+        ) -> DispatchResultWithPostInfo {
             // Ensure that only administrator entity can perform this administrative transaction
-            ensure!(Self::ensure_administrator(origin) == Ok(()), Error::<T>::MustBeAdministrator);
-
-            // Ensure that the pallet has not already been initialized 
-            ensure!(<Contributions<T>>::get() == None, Error::<T>::PalletAlreadyInitialized);
+            ensure!(
+                Self::ensure_administrator(origin) == Ok(()),
```
Comment (vedhavyas) [inline pallets/crowdloan-claim/src/lib.rs:553]: `is_ok` maybe ?

File: pallets/crowdloan-claim/src/lib.rs [hunk 8]
```diff
@@ -566,195 +450,234 @@ pub mod pallet {
...
+                        .priority(T::ClaimTransactionPriority::get())
+                        .and_provides((relaychain_account_id, parachain_account_id, claimed_amount))
+                        .longevity(
+                            TryInto::<u64>::try_into(T::ClaimTransactionLongevity::get())
+                                .unwrap_or(64_u64),
+                        )
+                        .propagate(true)
+                        .build();
+                } else {
+                    return InvalidTransaction::BadProof.into();
```
Comment (vedhavyas) [inline pallets/crowdloan-claim/src/lib.rs:630]: I think there can be `AlreadyClaimed` or `BadProof` depeneding on whiuch failed right?  maybe we can seperate those checks with respective errors ?

File: pallets/crowdloan-claim/src/lib.rs [hunk 9]
```diff
@@ -765,18 +688,52 @@ impl<T: Config> Pallet<T> {
...
+        // We could unwrap here, as we check in the calling function if module is initialized (i.e. if contributions is set)
+        // but better be safe than sorry...
+        let root = Self::contributions().ok_or(Error::<T>::PalletNotInitialized)?;
+
+        read_child_proof_check::<T::Hashing, Vec<Vec<u8>>>(root, proof, &child_info, keys).map_or(
+            Err(Error::<T>::InvalidProofOfContribution.into()),
+            |mut key_value_map| {
+                key_value_map
+                    .get_mut(&relay_account.encode())
+                    .map_or(&None, |map_val| map_val)
```
Comment (vedhavyas) [inline pallets/crowdloan-claim/src/lib.rs:720]: can't we return the Err here instead ?

**Review by vedhavyas — COMMENTED** [review]
minor nitpicks. Overall looks good

File: pallets/crowdloan-claim/src/lib.rs (hunk 2 above)
Comment (mustermeiszer) [reply pallets/crowdloan-claim/src/lib.rs:362]: Yes, this is better, will change that.

File: pallets/crowdloan-claim/src/lib.rs (hunk 8 above)
Comment (mustermeiszer) [reply pallets/crowdloan-claim/src/lib.rs:630]: Yes, good idea.

**Pull request merged**